Add contextual logger (log.FromContext)

xnyo opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana-plugin-sdk-go/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

4. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

-->

**What this PR does / why we need it**:

Adds a contextual logger, retrievable with `log.FromContext(ctx)`

This is a logger that's similar to `log.DefaultLogger`, but it's pre-populated with some contextual parameters:

- `pluginID`
- `endpoint`
- `traceID` (if available)
- `dsName` (if available)
- `dsUID` (if available)
- `uname` (if available)

Those are the same attributes logged in Grafana core's contextual logger:

https://github.com/grafana/grafana/blob/50504ba00803a291ea24e8a4a48b14494b97c744/pkg/services/pluginsintegration/clientmiddleware/contextual_logger_middleware.go#L27-L37

For this reason, this is especially useful for decoupling core plugins (replacing contextual logger with plugin sdk contextual logger).

This logger is added to the `context.Context` by the sdk adapter. This is a similar pattern to what happens in Grafana core (logging middleware).

Usage example:

```go
func (d *Datasource) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
	log.DefaultLogger.Info("no context", "a", "b")
	log.FromContext(ctx).Info("with context", "a", "b")
	// ...
```

```
INFO [10-17|17:23:56] no context                               logger=plugin.grafana-datasourcehttpbackend-datasource a=b
INFO [10-17|17:23:56] with context                             logger=plugin.grafana-datasourcehttpbackend-datasource endpoint=queryData uname=admin a=b dsName="Datasource Http Backend" dsUID=a7e_d3pVk pluginID=grafana-datasourcehttpbackend-datasource
```

**Which issue(s) this PR fixes**:

<!--

* Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #755

**Special notes for your reviewer**:

xnyo committed:
WIP contextual tracer

xnyo committed:
Add contextual logger for CheckHealth, CallResource and streaming

xnyo committed:
Add tests

xnyo committed:
lint

marefr reviewed (commented):
This looks like a very interesting approach. Great work. Just wonder if, lets say you have an existing logger created with logger.With() and you want to get a contextual logger from that then this approach wouldn't work. Not sure how common this would be though, but that would be how you would use a contextual logger in core grafana.

But maybe, rather than holding a logger in context holding attributes instead, similar to core grafana https://github.com/grafana/grafana/blob/8ac7736c9561d22789298582eb0cd90e3f789774/pkg/infra/log/log.go#L272-L300. WDYT? Lets discuss

xnyo committed:
PR review: store contextual attributes in context rather than logger

xnyo committed:
Update tests

xnyo commented:
Good point. I think it makes sense to make it similar to core Grafana. I have refactored it so it now stored the attributes in the context rather than the logger directly, and moved `FromContext` in the logger interface. This can now be used to create sub-loggers from existing ones:

```go
log.DefaultLogger.Info("default logger no context", "a", "b")

log.DefaultLogger.FromContext(ctx).Info("default contextual logger", "c", "d")

svcLogger := log.New().With("service", "myservice")
svcLogger.FromContext(ctx).Info("non-default contextual logger", "e", "f")
```

:point_down: 

```
INFO [10-18|13:02:35] default logger no context                logger=plugin.grafana-datasourcehttpbackend-datasource a=b
INFO [10-18|13:02:35] default contextual logger                logger=plugin.grafana-datasourcehttpbackend-datasource dsName="Datasource Http Backend" dsUID=a7e_d3pVk endpoint=queryData pluginID=grafana-datasourcehttpbackend-datasource c=d uname=admin
INFO [10-18|13:02:35] non-default contextual logger            logger=plugin.grafana-datasourcehttpbackend-datasource service=myservice dsName="Datasource Http Backend" dsUID=a7e_d3pVk pluginID=grafana-datasourcehttpbackend-datasource e=f endpoint=queryData uname=admin
```

Regarding `ContextualLogProviderFunc` which is present in core Grafana: I'm not sure, I haven't implemented it for now. It involves global variables and global state, I think keeping everything in context.Context is better. Also the approach that's implemented in this PR should be enough to handle most use cases. WDYT?

xnyo commented on backend/stream_adapter.go:
I decided to implement the contextual logger in the `stream_adapter` as well. We don't have it in core Grafana for some reason:

https://github.com/grafana/grafana/blob/50504ba00803a291ea24e8a4a48b14494b97c744/pkg/services/pluginsintegration/clientmiddleware/contextual_logger_middleware.go#L59C1-L69

I think it should also be the case for core Grafana, but maybe I am missing something. If there's a reason behind this decision, I will remove it from here. Otherwise I can also open a PR to add it to streaming methods in core Grafana as well

marefr replied:
It does make sense :+1: We should probably have it in core grafana as well

marefr commented on backend/log.go:
```diff
@@ -12,3 +26,22 @@ var Logger = log.DefaultLogger
 	return log.New().With(args...)
 }
+
+func withContextualLogAttributes(ctx context.Context, pCtx PluginContext, endpoint string) context.Context {
```
FYI for core plugins using the sdk logger we need to provide these contextual attributes as well together with existing https://github.com/grafana/grafana/blob/50504ba00803a291ea24e8a4a48b14494b97c744/pkg/services/pluginsintegration/clientmiddleware/contextual_logger_middleware.go#L59C1-L69 which currently is onyl for the grafana logger :) For now I guess we can duplicate this code to populate attributes for both core grafana logger and sdk logger in core grafana?

Eventually we probably want to include tenant id as well, but we can open a separate issue for that.

marefr reviewed (approved):
LGTM awesome work. Just added a comment, but not a blocker for these changes
